### PR TITLE
[ROCm] Added support for pytorch extensions to use HIP

### PR DESCRIPTION
**Pull request by ashishfarmer**
This pull request has changes for:
1. Enabling a torch module with HIP code to be compiled by cpp_extensions.py
2. Fixes for hipify module to be able to be used by a torch extension

cc: @ezyang @iotamudelta @jeffdaily

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 899d91fa:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=f6c46df856d0588360db5b807960d1fc5e888c36&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base f6c46df8 since Jan 27
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=899d91fa780d88074793e59aa0053d8b8a703d6c).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=91521571) from Jan 27 until Jan 27



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by ezyang** [comment]
@zou3519 Do you want to review this? You made major changes to cpp extensions recently.

File: torch/utils/cpp_extension.py
Comment (t-vi) [inline torch/utils/cpp_extension.py]: Can we please use `torch.version.hip` to check if we want to use ROCm?

**Comment by t-vi** [comment]
Awesome! I think the detection when to use HIP instead of CUDA needs fixing but other than that it looks reasonable at first sight.

**Comment by zou3519** [comment]
> @zou3519 Do you want to review this? You made major changes to cpp extensions recently.

Sure, I can review this.

At a glance I think this change is probably going to conflict a lot with https://github.com/pytorch/pytorch/pull/32495

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: I don't know much about how pytorch implements ROCM, so apologies if this is a silly question: can we use both AMD and nvidia gpus on the same machine, or is it mutually exclusive?

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: If possible we should enable some tests in `test_cpp_extension` to run in our ROCM CI (I assume they don't run right now).

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: (Putting this comment here because github won't let me put it where I want it)

It looks like this PR adds support for compiling ahead-of-time cpp extensions with ROCM. Did you also want to add support for just-in-time cpp extensions? I didn't see any changes to that setup. The code for that is roughly here: https://github.com/pytorch/pytorch/blob/e24ce0e524ba88df6ddab487ad9039ea1f324e01/torch/utils/cpp_extension.py#L1156-L1169

**Review by zou3519 — CHANGES_REQUESTED** [review]
We should try to add some tests if possible

**Comment by zou3519** [comment]
I'm not done reading through this yet, but I think it might be easier if we merge #32495 (which is still WIP by me) in first because that does a refactor of how cpp extensions do building. I'm happy to help fix up this PR to match the style of the refactor after that.

**Comment by ashishfarmer** [comment]
Thanks @zou3519 . We can hold this off until #32495 is merged. Both of the pull requests do touch the same sections of the code. We can refactor this PR once the other one is merged

**Comment by zou3519** [comment]
@ashishfarmer I've merged #32495. Here are some code pointers for what changed:
- There's a [new option](https://github.com/pytorch/pytorch/blob/ec1e9a1ae2cd53fa787beeef3a821a5b6b058322/torch/utils/cpp_extension.py#L322) to build ahead-of-time c++ extensions with ninja
- This is the [old option](https://github.com/pytorch/pytorch/blob/ec1e9a1ae2cd53fa787beeef3a821a5b6b058322/torch/utils/cpp_extension.py#L300) to build ahead-of-time c++ extensions with distutils.
- I reorganized the test files; [here is the test case for ahead-of-time c++ extensions](https://github.com/pytorch/pytorch/blob/master/test/test_cpp_extensions_aot.py), and [here is the test case for just-in-time c++ extensions](https://github.com/pytorch/pytorch/blob/master/test/test_cpp_extensions_jit.py)

Please let me know if there's anything I can help out with

**Comment by ashishfarmer** [comment]
Thank you @zou3519 for the heads up and the resources. I will incorporate my changes into the latest master with merged #32495 . Will reach out to you if I have any questions

File: torch/utils/cpp_extension.py
Comment (t-vi) [reply torch/utils/cpp_extension.py]: Currently we don't support AMD and NVidia in the same PyTorch.

File: torch/utils/cpp_extension.py
Comment (t-vi) [reply torch/utils/cpp_extension.py]: I think we want this eventually (and I briefly looked into it), but we could either include it in this PR or do it separately.

File: torch/utils/cpp_extension.py
Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: @zou3519 - I just removed `test_cpp_extensions_aot_no_ninja` from `ROCM_BLACKLIST`. However, the CI still skipped it. Is there anywhere else as well I should change?

File: torch/utils/cpp_extension.py
Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: In this PR, we enable ahead of time compilation for extensions using HIP with the legacy compilation way (not ninja). We will add just in time and ninja compilation separately

**Comment by ashishfarmer** [comment]
@zou3519 @t-vi - The PR is ready for review.

@ezyang - I did touch the hipify module to incorporate some changes and in https://github.com/pytorch/pytorch/pull/32669/commits/a235c4cc9c3057e87460a720427ca70941be377c it shows a way to first hipify and then create the extension. Let me know if you have any feedback on these changes or the approach

File: torch/utils/cpp_extension.py
Comment (zou3519) [reply torch/utils/cpp_extension.py]: Thanks for the clarification

**Comment by zou3519** [comment]
@ashishfarmer I'm not going to be able to get to this by the end of the day, but I'll definitely do a read-through on Tuesday

File: torch/utils/cpp_extension.py
Comment (zou3519) [reply torch/utils/cpp_extension.py]: I don't know how the mechanism works. @iotamudelta do you know how we can trigger the rocm tests on this PR?

File: test/cpp_extensions/setup.py
Comment (zou3519) [inline test/cpp_extensions/setup.py]: @ashishfarmer @t-vi do you know what our philosophy is around ROCm vs CUDA? Here I would prefer this to be called "ROCMExtension" (instead of CUDAExtension) so that there is never any ambiguity. But if the rest of pytorch implements ROCM through the CUDA API (i.e., is_cuda() returns True for a tensor on a ROCM device) then this seems OK to me.

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: Are .cu/.cuh files valid for ROCm?

File: test/cpp_extensions/setup.py
Comment (t-vi) [reply test/cpp_extensions/setup.py]: Yes, currently ROCm masquerades as CUDA for all PyTorch API purposes, so leaving this as CUDAExtension would seem consistent with what is done elsewhere. And really, if we ever make CUDA and ROCm available in the same PyTorch build, we'd probably want to make it a "GPUExtension" (named CUDAExtension, likely) and provide a list of backends this is compiled for.

File: torch/utils/cpp_extension.py
Comment (t-vi) [reply torch/utils/cpp_extension.py]: Yes. `hipify_python.hipify` (called here: https://github.com/pytorch/pytorch/pull/32669/files#diff-c87f1d9ef353fab9c4de839fe5adc8eeR33 ) will translate those from CUDA to HIP, to be consumed by the ROCm compiler.

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: Is there something else we can use to decide if we should be using to build with ROCM or not? Alternatively, creating a helper variable `is_rocm_extension` helps readability

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: I don't see this being used, it it necessary? We could also pass define macros like their command line arguments (`-DTHIS_MACRO`) so I'm not convinced that we should add a separate option for this.

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: It would be nice if we had some sort of

```
valid_files = ['.cu', '.cuh']
if rocm_extension:
  valid_files.append('.hip')
```
so that some eager dev doesn't come along and delete the '.hip' from here accidentally

**Comment by iotamudelta** [comment]
I see that @t-vi kindly has already answered some of the questions. In general, the difference CUDA - HIP/ROCm is small as possible. This means the frontend is shared and most of the backend is shared as well.

Concerning testing: ROCm test targets are accessible via pytorchbot.

File: torch/utils/cpp_extension.py
Comment (t-vi) [reply torch/utils/cpp_extension.py]: This is a deficiency (well, probably on purpose, but not good here) in the definition exclude_test that causes prefixes in the blacklist to match:
https://github.com/pytorch/pytorch/blob/44af8ee6cd3073ef2c6c6e8a61cfdde5dff9b345/test/run_test.py#L404
A quick fix could be to allow the blacklist to move to either move to regular expressions or allow "foo$" manually to mean "=='foo'" instead of "startswith(foo)".
Or one could rename the tests, no_ninja to the normal one and suffix the ninja one...

File: torch/utils/cpp_extension.py
Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: Done

File: torch/utils/cpp_extension.py
Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: Removed this unnecessary segment

File: torch/utils/cpp_extension.py
Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: done

File: torch/utils/cpp_extension.py
Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: @t-vi, That would be a change with implications throughout the pytorch test suite. Would you recommend not touching it for this pull request and look at it separately?

Comment (t-vi) [reply torch/utils/cpp_extension.py]: Upon further inspection, I'm not so sure whether the prefix matching even is the intended behaviour (e.g. non-x64 windows blacklists both cpp_extension_aot and cpp_extension_aot_no_ninja).
I think that:
- we really do want the test,
- renaming the ninja-enabled test to _ninja  that is the prefix is probably best and I would say it's legitimate to do it in this PR. Changing the overall behaviour less so.

Comment (t-vi) [reply torch/utils/cpp_extension.py]: Note that renaming the test also needs updating the windows blacklist etc.

Comment (zou3519) [reply torch/utils/cpp_extension.py]: Oh, so that's the problem. I agree with @t-vi that renaming the ninja-enabled test is the easiest thing to do for now. Long-term we should file an issue and figure out if the prefix matching behavior is intended behavior or not; afaict it isn't documented in the test runner.

Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: Thank you for the direction. Renamed the tests, the CI run should now have cpp_extensions test enabled for ROCm

File: test/run_test.py
Comment (zou3519) [inline test/run_test.py]: Oh, I see, you renamed test_cpp_extensions_aot -> test_cpp_extensions_aot_ninja and test_cpp_extensions_aot_no_ninja -> test_cpp_extensions_aot.

Could we please have the following instead:
- test_cpp_extensions_aot_ninja runs cpp extensions tests with ninja
- test_cpp_extensions_aot_no_ninja runs cpp extension tests without ninja
So that there is no fishy prefix behavior in run_test.py and to make it clear that the "no ninja build" is not the default one?

File: test/test_cpp_extensions_aot.py
Comment (zou3519) [inline test/test_cpp_extensions_aot.py]: Ditto, this should read either neutral or read as if "test_cpp_extensions_aot (with ninja)" is the default. Developers of features for cpp extensions should run test_cpp_extensions_aot (with ninja) because it is faster

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: nit: I think this is slightly easier to read if we have the following:
```
if _is_cuda_file(src):
  ...
elif isinstance(flags,dict):
  cflags = cflags['cxx']
if is_hip_extension:
   cflags = COMMON_HIPCC_FLAGS + cflags
```
so that we don't have to duplicate the `isinstance(cflags, dict)` check logic

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: nit: `is_hip_extension = (ROCM_HOME is not None) and (torch.version.hip is not None)` is sufficient. Also, global variables in this file have capital letters so we should probably stick with that style: IS_HIP_EXTENSION

**Review by zou3519 — COMMENTED** [review]
The contents of this look good to me. I think we should rename the test files so that there is no "bias" toward which test (test_cpp_extension_aot vs test_cpp_extension_aot_ninja) is the "default" one.

File: torch/utils/cpp_extension.py
Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: Simplified the condition

File: torch/utils/cpp_extension.py
Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: Done

File: test/run_test.py
Comment (zou3519) [reply test/run_test.py]: (resolved)

File: torch/utils/cpp_extension.py
Comment (zou3519) [inline torch/utils/cpp_extension.py]: (putting this here because github won't let me put it elsewhere)

We should toss a check for use_ninja to `BuildExtension.__init__`. If `use_ninja` is True and we are building a ROCm extension, we should error out gracefully so that ROCm users aren't confused about the extension not building. (I am not sure when we'll be able to build ROCm extensions with ninja; I know that's next on your list but this check is for just in case we don't get to that by the next release).

**Review by zou3519 — APPROVED** [review]
One last thing, otherwise lgtm

File: torch/utils/cpp_extension.py
Comment (ashishfarmer) [reply torch/utils/cpp_extension.py]: Thank you for the catch. Just added a fallback in the `__init__` similar to the behavior when ninja is not present in the system

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**
